test/network/BancorNetwork.ts

```diff
@@ -2117,7 +2117,7 @@ describe('BancorNetwork', () => {
             targetTokenAddress?: string;
         }
 
-        const tradeBySourceAmount = async (amount: BigNumberish, overrides: TradeOverrides = {}) => {
+        const tradeBySourceAmount = async (amount: BigNumberish, overrides: TradeOverrides = {}, getRetVal = false) => {
```

Comment:
1. `getRetVal` (terrible name) can be part of `TradeOverrides` (which can be renamed to `TradeOptions`). Its whole purpose is to group all additional parameters.

2. I don't like the style of passing something of "get ret val" to these functions. It's a strange interface. Let's always return the value in these cases (the performance cost is not that significant, since most of the time is wasted on setup anyway).

@yudilevi other ideas, perhaps?

Reply:
That was my initial solution.
However, we also need to be able to call this function when reverting, and then catch the revert and verify the error-message.
Up until now, we'd return the transaction promise, and resolve it by the caller, which would await for either successful completion, or a revert with an error.
Now we need to return two promises - one for the static-call, and the other one for the actual transaction (the same one which we've been returning up until now).
Our options for the first promise (the static-call one) are:
1. Await for (resolve) it before returning the result - that would be a problem, because it may revert.
2. Return it unresolved - that means that we will need to:
   - Make sure to resolve the two promises in the exact same order, in every test which checks both of them (because the 2nd promise changes the state and impact the result of the 1st promise)
   - Resolve each promise also in every test which doesn't make use of that promise (otherwise it's an unresolved promise "left hanging there", I'm not so sure what the consequences are for that); this also applies for tests which verify 'revert with an error', where we will need to expect the same revert on each one of the promises

Hence I chose a 3rd option:
Await for the 1st promise only when explicitly stated by the 'getRetVal' argument, which is set to true only in some specific tests where we're not catching a revert.

Reply:
That being said, the solution that I've employed is pretty horrific, so I have revised that into something much simpler, namely - each function can return with the transaction itself (as before), or a static-call of it.